### ISSUE TITLE
issue/confusion regarding ExecuteQuery

### ISSUE THREAD

**Issue by magicspon** [saiichihashimoto/sanity-typed#242]
Hello,

### Issue 3.  (comment edited)

```
export const postQuery = /* groq */ `*[_type == "post" && slug.current == $slug]{ title, slug }`

export type PostQuery = ExecuteQuery<
  typeof postQuery,
  { dataset: 'production'; _type: 'post' }
>
```
Incorrect types returned, getting `title: null; slug: null` rather than `title: string; slug: string`
This issue remains even with a super super simple schema 

i've also tried passing the string directly, i.e

```
export type PostQuery = ExecuteQuery<
  `*[_type == "post" && slug.current == $slug]{ title, slug }`,
  { dataset: 'production'; _type: 'post' }
>
```

Let me know if you want me to split these into separate tickets.  

Cheers

**Comment by saiichihashimoto** [comment]
Issue 1: I'll have to take a look, I'm out of town!
Issue 2: This may be out of my control, yes. You should be able to carry it ‘as‘ the native client type. I can make a method that converts it to feel more "safe".
Issue 3: Take a look at the ‘groq‘ package's README to see how to use this correctly, notably the ‘dataset‘, which will represent what type ‘*‘ will be in the query. The example you have will operate on the strong "production", which can't be filtered and projected.

Agreed this would've been better as separate tickets but probably fine for now.

**Comment by magicspon** [comment]
heyo,

issue 1: Thanks
issue 2: `client as SanityClient` does the job here, The error is replaced with the `Type instantiation is excessively deep and possibly infinite.ts(2589)` issue

issue 3. I've updated as per the docs, but i'm still seeing null for for title/slug and block.  Am I missing something?

```
export const postsQuery = /* groq */ `*[_type == 'post']{ title, slug, block }`

export type PostsQuery = ExecuteQuery<
  typeof postsQuery,
  { dataset: { _type: 'post' }[] }
>
```

**Comment by LBBO** [comment]
I'm seeing the third issue as well in a fresh project as the return value of `client.fetch(query)`. In case it's relevant, you can find the code [here](https://github.com/ory/summit-platform/pull/65)

**Comment by saiichihashimoto** [comment]
I'm out of town, so i haven't gotten to any of this. This definitely should've been three different issues, due to the conversation around them.

The third issue is working as expected. You're ultimately "operating" on the dataset. The `*` maps to the value of the `dataset` which, in your case, is defined as `{ _type: "post" }[]`. When you filter it by your filter, it remains unchanged. When you project `{ _type: "post" }` with any of your projections, you get issues. It has no title, slug, block, etc. What type should I return? How would I know the title is a string and not a number?

The point of the `groq` library is to give it the actual type of your dataset and a query, and then we can determine the type of the query's results. This library is used as an internal for the `client` library, so you can give it the schemas value types, make queries, and have them just work. I think the issue here is documentation.

**Comment by saiichihashimoto** [comment]
Awesome, will take a look at that one.

Either way, `{ _type: "post" }` is essentially your input type. You're projecting for certain fields off of it (`block`, `slug`, `title`), but it has none of those so it can't figure it out. You need to give it some value that has those fields, ie `{ _type: "post"; title: string; ... }`. Does that solve your issue @magicspon? I'm going to close this issue, feel free to keep commenting on it if you're still having issues.

@LBBO, I think your issue might be something different, but a 29 file PR is hard for me to parse through the issue. If it persists, let's make a new issue with a more specific description of your issue.

**Comment by magicspon** [comment]
@saiichihashimoto sorry still a bit confused.  Do you have a working example I could look at? 

![PixelSnap 2023-08-25 at 11 47 16@2x](https://github.com/saiichihashimoto/sanity-typed/assets/3268717/3f78c864-4330-41c9-b5f1-b9d38bcfe422)

![PixelSnap 2023-08-25 at 11 48 25@2x](https://github.com/saiichihashimoto/sanity-typed/assets/3268717/2d49a505-d7d7-4680-9283-6def214bb0f4)

**Comment by saiichihashimoto** [comment]
Can you create a minimal code snippet or repro I can look at? Screenshots aren't reproducible. It looks like `postsQuery` might be of the wrong type? I'm possible to tell like this.

**Comment by magicspon** [comment]
https://github.com/magicspon/sanity-typed-ts-error

**Comment by LBBO** [comment]
Playing around with @magicspon's repo in an effort to reproduce my issue there, I think I may have figured out a part of what the issue might be. The properties seem to be typed as `null` by `fetch` iff they aren't marked as required.

Consider this schema:
<img width="837" alt="image" src="https://github.com/saiichihashimoto/sanity-typed/assets/8076094/c415ac24-17ef-4c01-9936-58bd584146a5">

Notice how `name` is marked as required, which is reflected in the inferred schema values type (`name: string`) whereas `position` is not (inferred as `position?: string | undefined`). Querying that schema results in the following return type:

<img width="554" alt="image" src="https://github.com/saiichihashimoto/sanity-typed/assets/8076094/f0b2bbe3-4326-4329-a904-0ef9709ebf49">

Instead of receiving `position: null`, I would expect to receive `position?: string | undefined` here just as it's defined in the inferred schema value.

**Comment by saiichihashimoto** [comment]
Can you try this with `@sanity-typed/client` (and, if you're still using it, `@sanity-typed/groq`) at `1.1.3`? Let me know if it works! @magicspon @LBBO

**Comment by LBBO** [comment]
Nope, the `fetch` call now throws a `Type instantiation is excessively deep and possibly infinite .ts(2589)` type error. I tried creating a CodeSandbox to minimally reproduce the issue, but it's just not working at all.

**Comment by magicspon** [comment]
@saiichihashimoto no joy.

It's still no entirely clear to me how this should be used.  Could you add an example that has more that just the _type argument.

i presume i'm doing something wrong
![PixelSnap 2023-08-30 at 12 13 06](https://github.com/saiichihashimoto/sanity-typed/assets/3268717/2926f653-1da2-4dcc-abb9-258ab08510b6)

i'll update the example repo with your latest release

**Comment by saiichihashimoto** [comment]
I opened up this repo and... I'm running into such a massive slew of issues.

1.
<img width="301" alt="image" src="https://github.com/saiichihashimoto/sanity-typed/assets/2819256/df4aa2cc-d6fd-467d-b7c9-649a80f6344e">

no issues...

<img width="432" alt="image" src="https://github.com/saiichihashimoto/sanity-typed/assets/2819256/569f2d4a-7464-4eca-a662-948f216ce412">

issues? already?!

<img width="537" alt="image" src="https://github.com/saiichihashimoto/sanity-typed/assets/2819256/76a3844b-ed0e-4e0c-990d-56f09e8f2373">

obviously, no chance.

However, when I write it into a local test, there's no issues at all. So it's not a parsing/evaluating issue. For whatever reason, something that isn't an issue in my repo is an issue in yours. I think my way to test this is to create a minimal repo that imports the package and try out these queries. If it has an issue, then something between local testing and, building, releasing, installing, and importing is causing an issue. If the minimal repo doesn't have an issue, then there's something about your repo setup that causes the issue, and I'll have to do a bit of a repo-bisect to figure out what settings are required for this to work. I'm going to move this into #244 to have a cleaner place to track this.

2.
<img width="711" alt="image" src="https://github.com/saiichihashimoto/sanity-typed/assets/2819256/9965668d-987b-4b5e-8f5b-17598cbfb223">

Obviously, if the simpler query is failing, then this one definitely will.

<img width="588" alt="image" src="https://github.com/saiichihashimoto/sanity-typed/assets/2819256/7722df44-6e85-4d17-8bac-4d553208e8ed">
<img width="472" alt="image" src="https://github.com/saiichihashimoto/sanity-typed/assets/2819256/0769fbbe-e867-4575-bc2f-3177f2dc69e5">

However! A local test of this fails so, new issue!

<img width="638" alt="image" src="https://github.com/saiichihashimoto/sanity-typed/assets/2819256/30a20bb6-1ef8-4c45-80d9-dc5f1391fa30">

That `expr` looks suspicious, why isn't it expanding?

<img width="620" alt="image" src="https://github.com/saiichihashimoto/sanity-typed/assets/2819256/a0aa7a30-98cc-4c16-867c-54f1a3d2e448">

OK, let's poke around a bit. I'm guessing the `type` should be `Op`, right?

<img width="612" alt="image" src="https://github.com/saiichihashimoto/sanity-typed/assets/2819256/140c7ce6-c782-4f00-9c5c-32a9e7d9057b">

Oh.

<img width="611" alt="image" src="https://github.com/saiichihashimoto/sanity-typed/assets/2819256/68635c48-0f47-4491-bee0-9d92db2594cb">
<img width="609" alt="image" src="https://github.com/saiichihashimoto/sanity-typed/assets/2819256/fd3fa060-724d-4230-85a9-7d5a49ae320a">

So essentially, it's parsing `_type == "foo" && slug.current == $slug` as if it were all three of these:
1. `_type == ("foo" && slug.current == $slug)`
2. `(_type == "foo") && (slug.current == $slug)`
3. `(_type == "foo" && slug.current) == $slug`

This is ultimately what #219 is about. There's a chart of GROQ precedence that needs to be respected that I haven't even started to consider:

<img width="833" alt="image" src="https://github.com/saiichihashimoto/sanity-typed/assets/2819256/7a5d9e46-fdb3-4fd2-96de-e19e8975f750">

3.
I'd still say that you don't need to bother using `@sanity-typed/groq` directly. You should be using `@sanity-typed/client` directly like you are, giving it `SanityValues` and using `fetch` and having it magically work. That isn't working right now, hopefully because of these two issues, and solving them will hopefully solve your issue. Otherwise, you'd have to do something like this (which is what `@sanity-typed/client` is ultimately trying to accomplish for you):

<img width="496" alt="image" src="https://github.com/saiichihashimoto/sanity-typed/assets/2819256/a94e11ce-ace3-4807-b0eb-53a4491e5d18">

4.
For now... unfortunately, I think you'll just have to do `Extract`, `Pick`, and `Omit` magic against `SanityValues` to get what you want. Looks like `@sanity-typed/client` has some big issues here and it might be worth it to stick with `@sanity/client` until it works. These are some difficult issues and will take some time to nail down.

**Comment by saiichihashimoto** [comment]
Believe it or not, there's a fourth option for `_type == "foo" && slug.current == $slug`

4. `(((_type == "foo") && slug).current) == $slug`

🙄

**Comment by saiichihashimoto** [comment]
OK! I've solved the second issue. I know it won't look like it, because whatever the first issue is will say it's still too complex. And the first issue is being tracked in #244. At this point, I'm going to really close this issue. I know that your repo still doesn't work, It's still being worked on.

**Comment by magicspon** [comment]
@saiichihashimoto i'll have a look. Thanks for all your work on this. It sounds incredibly complicated!

**Comment by saiichihashimoto** [comment]
@LBBO I've identified the `null` value from a projection of an optional field and made a fix for it in #265
